\documentclass[10pt,a4paper,]{article}

%% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
%%   BScBI_CompGenomics_template.tex
%%
%%   A LaTeX template for MarkDown reports to submit as BScBI-CG practical exercises.
%%
%% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
%%                 CopyLeft 2020 (CC:BY-NC-SA) --- Josep F Abril
%%
%%   This file should be considered under the Creative Commons BY-NC-SA License
%%   (Attribution-Noncommercial-ShareAlike). The material is provided "AS IS", 
%%   mainly for teaching purposes, and is distributed in the hope that it will
%%   be useful, but WITHOUT ANY WARRANTY; without even the implied warranty
%%   of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
%%
%% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\usepackage[pdftex]{graphicx}
\usepackage{xcolor} %%% FIX %%%
\usepackage{fancyvrb}
\usepackage{comment}
\usepackage{fancyhdr} % Do not use \usepackage{fancybox} -> TOCs disappear
\usepackage{multirow}
\usepackage[nottoc,notlof,notlot,numbib]{tocbibind}

\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
\else % if luatex or xelatex
  \ifxetex
    \usepackage{mathspec}
    \usepackage{xltxtra,xunicode}
  \else
    \usepackage{fontspec}
  \fi
  \defaultfontfeatures{Mapping=tex-text,Scale=MatchLowercase}
  \newcommand{\euro}{â‚¬}
\fi
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
% use microtype if available
\IfFileExists{microtype.sty}{%
\usepackage{microtype}
\UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\usepackage[margin=1.5cm]{geometry}
\usepackage{natbib}
\bibliographystyle{plainnat}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\newenvironment{Shaded}{}{}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{#1}}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.49,0.56,0.16}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.53,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.56,0.13,0.00}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.73,0.13,0.13}{\textit{#1}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.02,0.16,0.49}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.40,0.40,0.40}{#1}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.74,0.48,0.00}{#1}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.73,0.40,0.53}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.10,0.09,0.49}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{#1}}}}
% .if(graphics).
% \usepackage{graphicx}
% \makeatletter
% \def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
% \def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
% \makeatother
% % Scale images if necessary, so that they will not overflow the page
% % margins by default, and it is still possible to overwrite the defaults
% % using explicit options in \includegraphics[width, height, ...]{}
% \setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% .endif.
\ifxetex
  \usepackage[setpagesize=false, % page size defined by xetex
              unicode=false, % unicode breaks when used with xetex
              xetex]{hyperref}
\else
  \usepackage[unicode=true]{hyperref}
\fi
\hypersetup{breaklinks=true,
            bookmarks=true,
            pdfauthor={Name SURNAME @ BScBI Computational Genomics},
            pdftitle={BScBI-CG Exercise 02},
            colorlinks=true,
            citecolor=green,
            urlcolor=blue,
            linkcolor=red,
            pdfborder={0 0 0}}
\urlstyle{same}  % don't use monospace font for urls
\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}
\setlength{\emergencystretch}{3em}  % prevent overfull lines
\setcounter{secnumdepth}{5}

%%% Compute current time using 24Hr. notation
{\count1=\time \divide\count1 by 60 \multiply\count1 by 60 %
 \count2=\time \advance\count2 by -\count1 %
 \divide\count1 by 60 %
 \xdef\currenttime{\the\count1:\ifnum\count2<10 0\fi\the\count2}}
\newcommand{\tstamp}{\textbf{\today\ /\ \currenttime}}

%%% Header/Footer Stuff
\def\thylab{\textbf{BSc BI --- Computational Genomics}}
\def\thylogo{\phantom{x}\raisebox{-2mm}{\includegraphics[height=6.5mm, width=6.5mm]{{docs/ESCI_logo}.png}}} %%% FIX %%%
%
\fancyhead{} % clear all fields
\fancyfoot{} % clear all fields
\fancyhead[RO,LE]{\thepage}
%\fancyhead[LO,RE]{\shorttit\quad\rightmark}
\fancyhead[LO]{\small{}BScBI-CG Exercise 02 Report}
\fancyhead[RE]{\small{}\rightmark}
\fancyfoot[LO,LE]{\small\textbf{\thylab}}
\fancyfoot[CO,CE]{\small\textsl{Cabello, A.}}
\fancyfoot[RO,RE]{\small\textbf{\today}\thylogo}
\renewcommand{\headrulewidth}{1pt}
\renewcommand{\footrulewidth}{1.5pt}

%.if(title).
%\title{.title..if(subtitle).\\\vspace{0.5em}{\large .subtitle.}.endif.}
%.endif.
\author{true}
\date{}

%% smaller font for code blocs...
\makeatletter
\newcommand{\verbatimfont}[1]{\renewcommand{\verbatim@font}{\ttfamily#1}}
\makeatother
\verbatimfont{\small}%

%% embedding external files
\newcommand{\loadfile}[3]{% BOXLABEL (escape "_") // ./relativepathtofile/FILENAME // REFLABEL
  \phantomsection\label{#3}\VerbatimInput[frame=single,framesep=5mm,fontsize=\small,label=\fbox{#1}]{#2}
}%

\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}

\def\CGVC{\href{https://aula.esci.upf.edu/course/view.php?id=5915}{Computational Genomics Virtual Campus at ESCI}}
\def\emboss{\texttt{EMBOSS}}
\def\Ecol{\textit{Escherichia coli}}
\def\ecol{\textit{E.~coli}}
\def\Bsub{\textit{Bacillus subtilis}}
\def\bsub{\textit{B.~subtilis}}
\def\Mgen{\textit{Mycoplasma genitalium}}
\def\mgen{\textit{M.~genitalium}}
\def\Mpne{\textit{Mycoplasma pneumoniae}}
\def\mpne{\textit{M.~pneumoniae}}
\def\gb{\textsc{GenBank}}
\def\GB{\href{https://www.ncbi.nlm.nih.gov/genome}{\gb}}

\begin{document}

%
\thispagestyle{empty}
%
\begin{titlepage}
\begin{center}
%
\ \vfill
%
\textbf{\Huge \textsc{BScBI-CG} \vskip 1.25ex \textsc{Practicals} \vskip 1.25ex
\textsc{Report}}\\[6ex]
%
\textbf{\Large%
\href{mailto:adrian.cabello@alum.esci.upf.edu?subject=[BScBI-CG Exercise 02]}{Adria Cabello}%
}\\%
\vskip 8ex
\textbf{\huge Exercise 02}
\vskip 8ex
\textbf{\large --- \today ---}
%
\end{center}
\vfill
%
%\begin{raggedright}
\hfill\begin{tabular}{r@{\hspace{1em}}l}
\includegraphics[height=2cm, width=4cm]{{docs/ESCI_logo}.png}
& %
\shortstack{%
\textbf{\textsc{\large Computational Genomics 2020--21}}\\[1ex]
\textsc{BSc in Bioinformatics}\\[1.25ex]
{\large \href{http://www.upf.edu/en/}{UPF} -- \href{http://www.fib.upc.edu/en/bioinformatica.html}{UPC} -- \href{http://www.ub.edu/web/ub/en/estudis/oferta_formativa/graus/fitxa/B/G1091/index.html?}{UB}}%
}%shortstack
\\[2ex]
\end{tabular}
%\end{raggedright}
%
\end{titlepage}
%
%\maketitle
\newpage
%thytitle

{
%\hypersetup{linkcolor=black}
\pagenumbering{roman}
\setcounter{page}{1}
\setcounter{tocdepth}{3}
\tableofcontents
%
\vfill
%\newpage
}
\listoftables
\vfill
\listoffigures
\vfill
%
\newpage

\pagenumbering{arabic}
\setcounter{page}{1}
\pagestyle{fancy}

\hypertarget{introduction}{%
\section{Introduction}\label{introduction}}

We want to analyze genome sequences of four bacteria species:
\textit{Escherichia coli}, \textit{Bacillus subtilis},
\textit{Mycoplasma genitalium}, and \textit{Mycoplasma pneumoniae}. All
the downstream commands from the initial template will focus on the
first of them, \textit{E.~coli}; you will need to perform similar
analyses for the other three species, then discuss differences among
those genomes from the data you will obtain.

\hypertarget{objectives}{%
\subsection{Objectives}\label{objectives}}

\begin{itemize}
\tightlist
\item
  To practice sequence retrieval commands and how to reformat records,
  for instance extracting \texttt{FASTA} records from a \texttt{GenBank}
  formatted file.
\item
  To implement and apply a running-windows approach to calculate
  sequence properties across a small set of genomic sequences.
\item
  To visualize those properties in order to compare the results obtained
  for the provided sequences.
\item
  To introduce \LaTeX~variables, item lists, and improved tabular
  environments.
\end{itemize}

\hypertarget{prerequisites}{%
\subsection{Prerequisites}\label{prerequisites}}

\hypertarget{installing-required-software}{%
\subsubsection{Installing required
software}\label{installing-required-software}}

As for the previous practical, we must ensure that at least
\texttt{pandoc} and \texttt{pdflatex} commands are running smoothly over
our report files. If you still need to install the base software, please
refer to \texttt{exercise\_00} and \texttt{exercise\_01}, as well as the
short tutorials from the
\href{https://aula.esci.upf.edu/course/view.php?id=5915}{Computational Genomics Virtual Campus at ESCI}.

For this practical you probably may need to install the following
packages:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#################################}
\CommentTok{# emboss - European molecular biology open software suite}

\CommentTok{# on a debian/ubuntu/mint linux system (DEBs)}
\ExtensionTok{apt-cache}\NormalTok{ search emboss     # to check if there is such a package}
\FunctionTok{sudo}\NormalTok{ apt-get install emboss }\CommentTok{# to install such a package}

\CommentTok{# on a redhat/fedora/centos linux system (RPMs)}
\ExtensionTok{yum}\NormalTok{ search emboss           # to check if there is such a package}
\FunctionTok{su}\NormalTok{ -c }\StringTok{'yum install emboss'}

\CommentTok{# on a SUSE/openSuse  linux system}
\ExtensionTok{zypper}\NormalTok{ search }\StringTok{"emboss"}
\FunctionTok{sudo}\NormalTok{ zypper install emboss}

\CommentTok{# on a Mac system using homebrew packages (**recommended option on a Mac**,}
\CommentTok{# see tutorial on the course introduction section materials at virtual campus)}
\ExtensionTok{brew}\NormalTok{ search emboss}
\CommentTok{# check the above command output, i.e. "brewsci/bio/emboss", to use on install:}
\FunctionTok{sudo}\NormalTok{ brew install brewsci/bio/emboss}

\CommentTok{# on a Mac system using anaconda packages (https://conda.io/docs/index.html)}
\ExtensionTok{conda}\NormalTok{ search emboss}
\CommentTok{# check the above command output to use on install:}
\FunctionTok{sudo}\NormalTok{ conda install -c bioconda emboss}

\CommentTok{# on a Mac system using mac ports (https://guide.macports.org/)}
\ExtensionTok{port}\NormalTok{ search emboss}
\CommentTok{# check the above command output to use on install:}
\FunctionTok{sudo}\NormalTok{ port install emboss}

\CommentTok{## IMPORTANT ## Do not mess your Mac system using all}
\CommentTok{#               of the previous three install options, use the one}
\CommentTok{#               already available on your system or install "homebrew".}

\CommentTok{# you can also install the package if available for the CygWin environment}
\CommentTok{# running on a Windows box (hhtp://www.cygwin.com/)}

\CommentTok{# add your packaging system here if you have not used any of the above commands...}
\end{Highlighting}
\end{Shaded}

From now on, we assume that you are using a Debian-based linux
distribution, so we will show only the corresponding set of commands for
that distribution.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#################################}
\CommentTok{# jellyfish - count k-mers in DNA sequences}
\FunctionTok{sudo}\NormalTok{ apt-get install jellyfish}
\end{Highlighting}
\end{Shaded}

\hypertarget{initializing-the-main-report-files}{%
\subsubsection{Initializing the main report
files}\label{initializing-the-main-report-files}}

As in the previous exercises, remember to download first the exercise
tarball from the
\href{https://aula.esci.upf.edu/course/view.php?id=5915}{Computational Genomics Virtual Campus at ESCI},
unpack this file, modify the files accordingly to the user within the
exercise folder, and set it as the current working directory for the
rest of the exercise\ldots{}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# You probably have already done this step.}
\FunctionTok{tar}\NormalTok{ -zxvf BScBI_CG2021_exercise_02.tgz}
\BuiltInTok{cd}\NormalTok{ exercise_02}

\CommentTok{# Rename report file including your "NAME" and "SURNAME"}
\FunctionTok{mv}\NormalTok{ -v README_BScBICG2021_exercise02_SURNAME_NAME.md \textbackslash{}}
\NormalTok{      README_BScBICG2021_exercise02_yourSurname_yourName.md}

\CommentTok{# Open exercise files using your text editor of choice}
\CommentTok{# (for instance vim, emacs, gedit, sublime, atom, ...);}
\CommentTok{# fix "NAME" and "SURNAME" placeholders on them}
\CommentTok{# and save those changes before continuing.}
\ExtensionTok{emacs}\NormalTok{  projectvars.sh \textbackslash{}}
\NormalTok{       README_BScBICG2021_exercise02_yourSurname_yourName.md }\KeywordTok{&}

\CommentTok{# Let's start with some initialization.}
\BuiltInTok{source}\NormalTok{ projectvars.sh}
\BuiltInTok{echo} \VariableTok{$WDR}

\CommentTok{# Once you have run the commands that are already in the initial}
\CommentTok{# MarkDown document, you are probably ready to run this:}
\ExtensionTok{runpandoc}
\end{Highlighting}
\end{Shaded}

Let's start with the analyses, and may the shell be with you\ldots{}

\newpage

\hypertarget{calculating-genome-sequence-properties}{%
\section{Calculating Genome Sequence
Properties}\label{calculating-genome-sequence-properties}}

\hypertarget{datasets}{%
\subsection{Datasets}\label{datasets}}

\input{docs/tbl_genbank_summary_info_genomes}

Table \ref{tbl:genbankgenomesinfo} provides an overview of the four
bacterial genomes we have to analyze on this exercise, for which we
provide a short description here:.

\textbackslash{}begin\{itemize\} \% IMPORTANT: \% do not use MarkDown
marks inside a LaTeX block, as they will not be processed by pandoc

\item

\textbackslash{}href\{https://www.ncbi.nlm.nih.gov/genome/?term=escherichia\%20coli\}\{\textit{E.~coli}\}
is typically present in the lower intestine of humans; is easily grown
in a laboratory setting and also readily amenable to genetic
manipulation, making it one of the most studied prokaryotic model
organisms. We will work with this species representative genome, which
is
\href{https://www.ncbi.nlm.nih.gov/genome/167?genome_assembly_id=161521}{\textit{E.~coli}\ strain K-12 substr. MG1655}.

\item

\href{https://www.ncbi.nlm.nih.gov/genome/?term=bacillus%20subtilis}{\bsub} is a model organism for prokaryotic cell differentiation and development, and was one of the first bacteria studied. Representative genome for this species is \href{https://www.ncbi.nlm.nih.gov/genome/665?genome_assembly_id=300274}{\bsub\ subsp. \textit{subtilis} strain 168}.
\item Mycoplasmas carry the smallest genomes of self-replicating cells together with the smallest set of functional coding regions; \href{https://www.ncbi.nlm.nih.gov/genome/?term=mycoplasma+genitalium}{\textit{Mycoplasma genitalium}} genome was the second to be reported in 1995\footnote{"The minimal gene complement of \textit{Mycoplasma genitalium}". Fraser CM, et al. \textit{Science}, 1995.}. The representative genome is \href{https://www.ncbi.nlm.nih.gov/genome/474?genome_assembly_id=300158}{\textit{M.~genitalium}\ G37}.
\item \href{https://www.ncbi.nlm.nih.gov/genome/?term=mycoplasma+pneumoniae}{\textit{M.~pneumoniae}} causes respiratory tract infections. We are going to use \href{https://www.ncbi.nlm.nih.gov/genome/1028?genome_assembly_id=300492}{\textit{M.~pneumoniae}\ M129} as representative genome.
\end{itemize}

It's time to get the sequences from a set of links we have retrieved from \href{https://www.ncbi.nlm.nih.gov/genome}{\textsc{GenBank}}\ genome division. We are not going to take just the sequences in `fasta` format, we will download them in \textsc{GenBank}\ format this time.

```{.sh}
# IMPORTANT: ensure that your WDR variable definition in projectvars.sh
#            does not contain a path having white-spaces on the folder names.

export DT=$WDR/data
mkdir -v $DT
# You can also add the previous var definition to your 'projectvars.sh' file
# so it will be saved and can be easily reused when sourcing the file again.

# Downloading the Ecol genome in GenBank format
GBFTP=ftp://ftp.ncbi.nlm.nih.gov/genomes/all

wget $GBFTP/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.gbff.gz \
     -O $DT/Ecol_referencegenome.gb.gz

# the other three genomes are available through the following paths:
#
#  Bsub GCF/000/009/045/GCF_000009045.1_ASM904v1/GCF_000009045.1_ASM904v1_genomic.gbff.gz
#  Mgen GCF/000/027/325/GCF_000027325.1_ASM2732v1/GCF_000027325.1_ASM2732v1_genomic.gbff.gz
#  Mpne GCF/000/027/345/GCF_000027345.1_ASM2734v1/GCF_000027345.1_ASM2734v1_genomic.gbff.gz
#
# save them as Bsub_referencegenome.gb.gz, Mgen_referencegenome.gb.gz,
# and Mpne_referencegenome.gb.gz respectively.
# For such a task, you can use a shell loop for instance:

while read Ospc Gftp;
  do {
    echo "# Downloading genome sequence for $Ospc" 1>&2;
    wget $GBFTP/${Gftp}_genomic.gbff.gz \
         -O $DT/${Ospc}_referencegenome.gb.gz
  }; done <<'EOF'
Bsub GCF/000/009/045/GCF_000009045.1_ASM904v1/GCF_000009045.1_ASM904v1
Mgen GCF/000/027/325/GCF_000027325.1_ASM2732v1/GCF_000027325.1_ASM2732v1
Mpne GCF/000/027/345/GCF_000027345.1_ASM2734v1/GCF_000027345.1_ASM2734v1
EOF

### IMPORTANT NOTE ###
#
# If the firewall does not allow you to connect to the original NCBI ftp site
# then you can run the following commands to download files from
# the https alternate repository at compgen.bio.ub.edu server.
# Just remind to replace the user and password strings with those
# from the slides for the introduction to the practicals.
#
GBFTP=https://compgen.bio.ub.edu/~jabril/teaching/BScBI-CG2021/repo_ex2

while read Ospc Gftp;
  do {
    echo "# Downloading genome sequence for $Ospc" 1>&2;
    wget --user="UUUUUUUUUUUU" \
         --password="XXXXXXXX" \
         $GBFTP/${Ospc}_referencegenome.gb.gz \
         -O $DT/${Ospc}_referencegenome.gb.gz
  }; done <<'EOF'
Ecol GCF_000005845.2_ASM584v2
Bsub GCF_000009045.1_ASM904v1
Mgen GCF_000027325.1_ASM2732v1
Mpne GCF_000027345.1_ASM2734v1
EOF
```


## Retrieving the sequences

Let's extract the raw genomic sequences from the \textsc{GenBank}\ formated files:


```{.sh}
# for manual pages on this emboss tool run:  tfm seqret
#
SPC="Ecol"

zcat $DT/${SPC}_referencegenome.gb.gz | \
  seqret -sequence genbank::stdin -outseq fasta::stdout | \
    gzip -9c - > $DT/${SPC}_referencegenome.fa.gz

# let's verify if fasta sequence has same length as reported in the GenBank file

zegrep '^LOCUS'  $DT/${SPC}_referencegenome.gb.gz
#> LOCUS       NC_000913            4641652 bp    DNA     circular CON 08-AUG-2016

zcat $DT/${SPC}_referencegenome.fa.gz | \
  infoseq -sequence fasta::stdin \
          -noheading -only -name -length -pgc
#> Display basic information about sequences
#> NC_000913      4641652 50.79

### repeat the commands for the other three genomes ###
```

From the output of the two commands, we can conclude that fasta sequence for the downloaded \textit{E.~coli}\ genome has the correct length, 4641652bp, and that the GC content is almost the same as the one reported on Table \ref{tbl:genbankgenomesinfo}, 50.79\% versus 50.8\% respectively (so the difference is due to rounding to one decimal position).


## Summary of sequence content

### Chaos-plot

\texttt{EMBOSS}\ suite has a command to calculate [chaos plots](http://emboss.sourceforge.net/apps/cvs/emboss/apps/chaos.html), a simple graphical representation of sequence composition that we can use to visually compare the four genomes analyzed on this exercise.

```{.sh}
zcat $DT/${SPC}_referencegenome.fa.gz | \
  chaos -sequence fasta::stdin -verbose \
        -graph png -gtitle "$SPC chaos plot" \
        -goutfile $WDR/images/${SPC}_chaosplot

### repeat the commands for the other three genomes ###
```

You __must__ include here a \LaTeX\ figure, defined as a table of two rows and two columns containing the four `png` plots, using `input` to load an external `tex` file stored in the `docs` directory (we had already examples on the previous exercise, see for instance "`exercise_01/docs/fig_histograms.tex`").


### Computing GC content variation across the genome

We will use a short script in `Perl` to calculate a parameter over the genome, GC content for instance, using a running window. You can try to rewrite in `Python` or any other programming language, but we should focus here on the effect of the window size on the final results. We define a short script in the following code chunk that you must copy into a file within the bin folder. It takes two parameters, the window length and the input sequence, so that you can play with several lengths to evaluate which one can provide the best comparison across genomes. Once you choose a window length on the \textit{Escherichia coli}\ genome, you can run the same command fixing that parameter and changing the input file for the other three genomes.

```{.perl}
#!/usr/bin/perl
#
# you can save this script as "bin/genomicgcwindows.pl"
#
use strict;
use warnings;

# variables initialization
my $window = shift @ARGV;
$window < 10 && die("# ERROR: window length must be a positive integer equal or greater than 10\n");
my $step = int($window / 2); # we have chosen to fix this parameter
my %SEQS = (); # just in case there is more than one sequence on the input

# read sequences
my $sid = undef;
while (<>) {
    next if /^\s*$/o;
    chomp;
    $_ =~ /^>/ && do {  # finding the sequence header with its name
        ($sid, undef) = split /\s+/, $_;
        exists($SEQS{$sid}) || ($SEQS{$sid} = '');
        next;
    };
    defined($sid) || next;
    $_ =~ s/\s+//og; # just in case there are white spaces on the sequence
    $SEQS{$sid} .= uc($_);
}; # while $_

# analyze sequences
foreach my $sid (keys %SEQS) {
    my $seq = $SEQS{$sid};
    for (my $n = 0; $n < length($seq) - $window + 1; $n += $step) {
        my $winseq = substr($seq, $n, $window);
        printf "%s %d %.1f\n", $sid, $n + $step, &getGC(\$winseq,$window);
    };
}; \# foreach \$sid

exit(0);

\hypertarget{available-functions}{%
\section{available functions}\label{available-functions}}

sub getGC() \{ my (\$sq, \$wn) = \citet{_}; my \$gc = 0; for (my \$c =
0; \$c \textless{} \$wn; \$c++) \{ \(gc++ if substr(\)\$sq, \$c, 1)
=\textasciitilde{} /{[}GC{]}/o; \}; \# for \$c return \$gc / \$wn * 100;
\} \# getGC ```

Let's run the \texttt{Perl} scrip on a set of increasing windows
lengths.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# provide execution permissions to the perl script}
\FunctionTok{chmod}\NormalTok{ a+x }\VariableTok{$WDR}\NormalTok{/bin/genomicgcwindows.pl}

\CommentTok{# running on Ecoli genome sequence}

\KeywordTok{for} \ExtensionTok{WNDW}\NormalTok{ in 100 200 500 1000 2000 5000 10000}\KeywordTok{;}
  \KeywordTok{do} \KeywordTok{\{}
    \BuiltInTok{echo} \StringTok{"# Running windowed GC analysis on }\VariableTok{$SPC}\StringTok{ for window length = }\VariableTok{$WNDW}\StringTok{"} \OperatorTok{1>&2}\KeywordTok{;}
    \FunctionTok{zcat} \VariableTok{$DT}\NormalTok{/}\VariableTok{$\{SPC\}}\NormalTok{_referencegenome.fa.gz }\KeywordTok{|} \KeywordTok{\textbackslash{}}
      \VariableTok{$WDR}\ExtensionTok{/bin/genomicgcwindows.pl} \VariableTok{$WNDW}\NormalTok{ - }\KeywordTok{|} \KeywordTok{\textbackslash{}}
         \FunctionTok{gzip}\NormalTok{ -c9 - }\OperatorTok{>} \VariableTok{$WDR}\NormalTok{/stats/}\VariableTok{$\{SPC\}}\NormalTok{_genomegcanalysis_wlen}\VariableTok{$WNDW}\NormalTok{.tbl.gz}\KeywordTok{;}
  \KeywordTok{\}}\NormalTok{; }\KeywordTok{done}\NormalTok{;}

\CommentTok{# just check the output}
\FunctionTok{ls}\NormalTok{ -1 }\VariableTok{$WDR}\NormalTok{/stats/}\VariableTok{$\{SPC\}}\NormalTok{_genomegcanalysis_wlen*.tbl.gz }\KeywordTok{|} \KeywordTok{\textbackslash{}}
  \KeywordTok{while} \BuiltInTok{read} \VariableTok{FL}\NormalTok{;}
    \KeywordTok{do} \KeywordTok{\{}
      \BuiltInTok{echo} \VariableTok{$FL}\KeywordTok{;}
      \FunctionTok{zcat} \VariableTok{$FL} \KeywordTok{|} \FunctionTok{head}\NormalTok{ -2}\KeywordTok{;}
    \KeywordTok{\}}\NormalTok{; }\KeywordTok{done}\NormalTok{;}
\CommentTok{#> stats/Ecol_genomegcanalysis_wlen100.tbl}
\CommentTok{#> >NC_000913 50 42.0}
\CommentTok{#> >NC_000913 100 34.0}
\CommentTok{#> stats/Ecol_genomegcanalysis_wlen200.tbl}
\CommentTok{#> >NC_000913 100 37.0}
\CommentTok{#> >NC_000913 200 44.0}
\CommentTok{#> stats/Ecol_genomegcanalysis_wlen500.tbl}
\CommentTok{#> >NC_000913 250 46.4}
\CommentTok{#> >NC_000913 500 52.8}
\CommentTok{#> stats/Ecol_genomegcanalysis_wlen1000.tbl}
\CommentTok{#> >NC_000913 500 50.7}
\CommentTok{#> >NC_000913 1000 54.4}
\CommentTok{#> stats/Ecol_genomegcanalysis_wlen2000.tbl}
\CommentTok{#> >NC_000913 1000 51.9}
\CommentTok{#> >NC_000913 2000 52.5}
\CommentTok{#> stats/Ecol_genomegcanalysis_wlen5000.tbl}
\CommentTok{#> >NC_000913 2500 53.0}
\CommentTok{#> >NC_000913 5000 52.2}
\CommentTok{#> stats/Ecol_genomegcanalysis_wlen10000.tbl}
\CommentTok{#> >NC_000913 5000 52.1}
\CommentTok{#> >NC_000913 10000 50.8}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{### repeat the commands for the other three genomes }\AlertTok{###}
\end{Highlighting}
\end{Shaded}

We can plot each of those tables using the nucleotide positions on the
X-axis and the computed GC content as Y-axes, those figures should be
five times wider than taller that will allow us to stack them for
comparing the results of the different window lengths.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{R}
\CommentTok{# then assuming you use R command-line shell from the terminal... ;^D}

\CommentTok{# example here for Ecol and window length equal to 100bp}

\NormalTok{GC_avg <-}\StringTok{ }\FloatTok{50.79}\NormalTok{; }\CommentTok{# the whole genome average GC content}

\NormalTok{ZZ <-}\StringTok{ }\KeywordTok{gzfile}\NormalTok{(}\StringTok{'stats/Ecol_genomegcanalysis_wlen100.tbl.gz'}\NormalTok{);}
\NormalTok{GC_w100 <-}\StringTok{ }\KeywordTok{read.table}\NormalTok{(ZZ, }\DataTypeTok{header=}\OtherTok{FALSE}\NormalTok{);}
\KeywordTok{colnames}\NormalTok{(GC_w100) <-}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"CHRid"}\NormalTok{,}\StringTok{"NUCpos"}\NormalTok{,}\StringTok{"GCpct"}\NormalTok{);}

\KeywordTok{summary}\NormalTok{(GC_w100)}
\CommentTok{#>        CHRid           NUCpos            GCpct      }
\CommentTok{#>  NC_000913:92832   Min.   :     50   Min.   :15.00  }
\CommentTok{#>                    1st Qu.:1160438   1st Qu.:47.00  }
\CommentTok{#>                    Median :2320825   Median :52.00  }
\CommentTok{#>                    Mean   :2320825   Mean   :50.79 (*)}
\CommentTok{#>                    3rd Qu.:3481212   3rd Qu.:56.00  }
\CommentTok{#>                    Max.   :4641600   Max.   :78.00}
\CommentTok{# mean of all GCpct (*) should be closer to the whole genome average GC, should it?}

\KeywordTok{library}\NormalTok{(ggplot2);}

\NormalTok{G <-}\StringTok{ }\KeywordTok{ggplot}\NormalTok{(GC_w100, }\KeywordTok{aes}\NormalTok{(}\DataTypeTok{x=}\NormalTok{NUCpos, }\DataTypeTok{y=}\NormalTok{GCpct)) }\OperatorTok{+}
\StringTok{        }\KeywordTok{geom_line}\NormalTok{(}\DataTypeTok{colour =} \StringTok{"blue"}\NormalTok{) }\OperatorTok{+}
\StringTok{        }\KeywordTok{theme_bw}\NormalTok{() }\OperatorTok{+}
\StringTok{        }\KeywordTok{geom_hline}\NormalTok{(}\DataTypeTok{yintercept=}\NormalTok{GC_avg, }\DataTypeTok{colour=}\StringTok{"red"}\NormalTok{, }\DataTypeTok{linetype=}\StringTok{"dashed"}\NormalTok{, }\DataTypeTok{size=}\FloatTok{1.5}\NormalTok{) }\OperatorTok{+}
\StringTok{        }\KeywordTok{ggtitle}\NormalTok{(}\StringTok{"E.coli GC content over the genome (window length = 100bp)"}\NormalTok{) }\OperatorTok{+}
\StringTok{        }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{x=}\StringTok{"Genomic Coords (bp)"}\NormalTok{, }\DataTypeTok{y=}\StringTok{"%GC Content"}\NormalTok{);}

\KeywordTok{ggsave}\NormalTok{(}\StringTok{"images/Ecol_genomegcanalysis_wlen100.png"}\NormalTok{,}
       \DataTypeTok{plot=}\NormalTok{G, }\DataTypeTok{width=}\DecValTok{25}\NormalTok{, }\DataTypeTok{height=}\DecValTok{8}\NormalTok{, }\DataTypeTok{units=}\StringTok{"cm"}\NormalTok{, }\DataTypeTok{dpi=}\DecValTok{600}\NormalTok{);}
\end{Highlighting}
\end{Shaded}

Include here a figure combining the plots for the set of window lengths
(100, 200, 500, 1000, 2000, 5000, and 10000). Then choose one of those
windows lengths and provide the commands to analyze the other three
genomic sequences. After that, you can include another figure stacking
the results for that window length on all the genomes.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{### repeat the commands for the other three genomes }\AlertTok{###}
\end{Highlighting}
\end{Shaded}

\hypertarget{analysis-of-k-mer-composition}{%
\subsection{\texorpdfstring{Analysis of \emph{k}-mer
composition}{Analysis of k-mer composition}}\label{analysis-of-k-mer-composition}}

There are many software tools to account for the \emph{k}-mers appearing
in a genomic sequence, we will use \texttt{jellyfish} for this purpose.
It produces a summary file where we can easily get the number of total,
distinct and unique \emph{k}-mers. We can even compare which
\emph{k}-mers appear in more than one species genome, but we will focus
this analysis on those numbers from the summary file.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{zcat} \VariableTok{$DT}\NormalTok{/}\VariableTok{$\{SPC\}}\NormalTok{_referencegenome.fa.gz }\KeywordTok{|} \KeywordTok{\textbackslash{}}
    \ExtensionTok{jellyfish}\NormalTok{ count -m 20 -C -t 4 -c 8 -s 10000000 /dev/fd/0 \textbackslash{}}
\NormalTok{              -o }\VariableTok{$WDR}\NormalTok{/stats/}\VariableTok{$\{SPC\}}\NormalTok{_jellyfish_k20.counts}\KeywordTok{;}

\ExtensionTok{jellyfish}\NormalTok{ stats }\VariableTok{$WDR}\NormalTok{/stats/}\VariableTok{$\{SPC\}}\NormalTok{_jellyfish_k20.counts}\KeywordTok{;}
\CommentTok{#> Unique:    4507760}
\CommentTok{#> Distinct:  4542190}
\CommentTok{#> Total:     4641633}
\CommentTok{#> Max_count:      82}
\CommentTok{# also consider that the total theoretical sequences of k=20 is 4^20 = 1,099511628e+12}
\end{Highlighting}
\end{Shaded}

We can also combine commands into a shell function, the code below runs
the same two \texttt{jellyfish} commands of the previous code block:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# we define here the shell function}
\KeywordTok{function}\FunctionTok{ jellyfish_on_kmer ()} \KeywordTok{\{}
  \VariableTok{THYSPC=$1}\NormalTok{;}
  \VariableTok{KMERSZ=$2}\NormalTok{;}
  \BuiltInTok{echo} \StringTok{"# }\VariableTok{$\{THYSPC\}}\StringTok{ - }\VariableTok{$\{KMERSZ\}}\StringTok{"} \OperatorTok{1>&2}\KeywordTok{;}
  \FunctionTok{zcat} \VariableTok{$DT}\NormalTok{/}\VariableTok{$\{THYSPC\}}\NormalTok{_referencegenome.fa.gz }\KeywordTok{|} \KeywordTok{\textbackslash{}}
    \ExtensionTok{jellyfish}\NormalTok{ count -m }\VariableTok{$KMERSZ}\NormalTok{ -C -t 4 -c 8 -s 10000000 /dev/fd/0 \textbackslash{}}
\NormalTok{                -o }\VariableTok{$WDR}\NormalTok{/stats/}\VariableTok{$\{SPC\}}\NormalTok{_jellyfish_k}\VariableTok{$\{KMERSZ\}}\NormalTok{.counts}\KeywordTok{;}         
  \ExtensionTok{jellyfish}\NormalTok{ stats }\VariableTok{$WDR}\NormalTok{/stats/}\VariableTok{$\{SPC\}}\NormalTok{_jellyfish_k}\VariableTok{$\{KMERSZ\}}\NormalTok{.counts}\KeywordTok{;}
\KeywordTok{\}}

\CommentTok{# here we use the previous function on a shell command-line,}
\CommentTok{#   with different parameters}
\CommentTok{#}
\ExtensionTok{jellyfish_on_kmer}\NormalTok{ Ecol 20}\KeywordTok{;}
\CommentTok{# # Ecol - 20}
\CommentTok{# Unique:    4507760}
\CommentTok{# Distinct:  4542190}
\CommentTok{# Total:     4641633}
\CommentTok{# Max_count: 82}
\CommentTok{#}
\ExtensionTok{jellyfish_on_kmer}\NormalTok{ Bsub 35}\KeywordTok{;}
\CommentTok{# # Bsub - 35}
\CommentTok{# Unique:    4152426}
\CommentTok{# Distinct:  4164735}
\CommentTok{# Total:     4215572}
\CommentTok{# Max_count: 10}
\CommentTok{#}
\CommentTok{#   or within loops...}
\CommentTok{#}
\KeywordTok{for} \ExtensionTok{SPC}\NormalTok{ in Ecol Bsub}\KeywordTok{;}
  \KeywordTok{do} \KeywordTok{\{}
    \KeywordTok{for} \ExtensionTok{KSZ}\NormalTok{ in 10 15 20}\KeywordTok{;}
      \KeywordTok{do} \KeywordTok{\{}
       \ExtensionTok{jellyfish_on_kmer} \VariableTok{$SPC} \VariableTok{$KSZ}\KeywordTok{;}
      \KeywordTok{\}}\NormalTok{; }\KeywordTok{done}\NormalTok{;}
  \KeywordTok{\}}\NormalTok{; }\KeywordTok{done}\NormalTok{;}
\CommentTok{# ...}

\CommentTok{# yet another option is to move the shell commands}
\CommentTok{# that perform single tasks into shell scripts instead of functions.}
\end{Highlighting}
\end{Shaded}

Try different \emph{k}-mer sizes (i.e.~10, 15, 20, 25, 30, 35, and 40),
on the genomic sequences of the four species and summarize them into
another \LaTeX~table to include below. You can take
``\texttt{docs/tbl\_genbank\_summary\_info\_genomes.tex}'' as example to
create this table.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{### repeat the commands for the other three genomes }\AlertTok{###}
\end{Highlighting}
\end{Shaded}

\hypertarget{discussion}{%
\section{Discussion}\label{discussion}}

\textbf{IMPORTANT} Discuss your results here (around 300 words). And
remember to include in the Appendices section (see page
\pageref{sec:appendices}), any extra script you wrote from this exercise
\texttt{bin} folder using the \texttt{loadfile} macro. We can take
advantage of the \LaTeX~referencing capabilities, as described in the
first exercise template.

\clearpage

\hypertarget{appendices}{%
\section{Appendices}\label{appendices}}

\label{sec:appendices}

\hypertarget{software}{%
\subsection{Software}\label{software}}

We have used the following versions:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{uname}\NormalTok{ -a}
\CommentTok{# Linux aleph 4.15.0-118-generic #119-Ubuntu SMP}
\CommentTok{# Tue Sep 8 12:30:01 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux}

\FunctionTok{wget}\NormalTok{ --version}
\CommentTok{# GNU Wget 1.20.1 built on linux-gnu.}

\ExtensionTok{infoseq}\NormalTok{ -version}
\CommentTok{# EMBOSS:6.6.0.0}

\ExtensionTok{jellyfish}\NormalTok{ -V}
\CommentTok{# jellyfish 2.2.8}

\ExtensionTok{R}\NormalTok{ --version}
\CommentTok{# R version 3.5.2 (2018-12-20) -- "Eggshell Igloo"}
\CommentTok{# Copyright (C) 2018 The R Foundation for Statistical Computing}
\CommentTok{# Platform: x86_64-pc-linux-gnu (64-bit)}
\end{Highlighting}
\end{Shaded}

\hypertarget{supplementary-files}{%
\subsection{Supplementary files}\label{supplementary-files}}

\label{sec:supplfiles}

\hypertarget{project-specific-scripts}{%
\subsubsection{Project specific
scripts}\label{project-specific-scripts}}

\loadfile{an\_script\_example.pl}{bin/an_script_example.pl}{prg:scriptexamplePERL}

\hypertarget{shell-global-vars-and-settings-for-this-project}{%
\subsubsection{Shell global vars and settings for this
project}\label{shell-global-vars-and-settings-for-this-project}}

\loadfile{projectvars.sh}{projectvars.sh}{prg:projectvarsBASH}

\hypertarget{about-this-document}{%
\subsection{About this document}\label{about-this-document}}

This document was be compiled into a PDF using \texttt{pandoc} (see
\texttt{projectvars.sh} from previous subsection) and some
\texttt{LaTeX} packages installed in this linux system.
\texttt{synaptic}, \texttt{apt-get} or \texttt{aptitude} can be used to
retrieve and install those tools from linux repositories. As the
\texttt{raw\_tex} extension has been provided to the
\texttt{markdown\_github} and \texttt{tex\_math\_dollars} formats, now
this document supports inline \LaTeX~and inline formulas!

You can get further information from the following links about the
\href{http://daringfireball.net/projects/markdown/syntax\#link}{Mark
Down syntax}, as well as from the manual pages (just type
\texttt{man\ pandoc} and/or \texttt{man\ pandoc\_markdown}).


\end{document}
